pkg/services/cleanup/cleanup.go

```diff
@@ -57,66 +62,97 @@ type CleanUpService struct {
-				time.Minute*10, func(context.Context) {
-					srv.deleteOldLoginAttempts(ctx)
-				})
-			if err != nil {
-				srv.log.Error("failed to lock and execute cleanup of old login attempts", "error", err)
-			}
+			srv.clean(ctx)
 		case <-ctx.Done():
 			return ctx.Err()
 		}
 	}
 }
 
-func (srv *CleanUpService) cleanUpOldAnnotations(ctx context.Context) {
+func (srv *CleanUpService) clean(ctx context.Context) {
+	const timeout = 9 * time.Minute
+	start := time.Now()
+	ctx, span := srv.tracer.Start(ctx, "cleanup background job")
+	defer span.End()
+	tID := tracing.TraceIDFromContext(ctx, false)
+	logger := srv.log.New("traceID", tID)
+	ctx, cancelFn := context.WithTimeout(ctx, timeout)
+	defer cancelFn()
+
+	cleanupJobs := []cleanUpJob{
+		{"clean up temporary files", srv.cleanUpTmpFiles},
+		{"delete expired snapshots", srv.deleteExpiredSnapshots},
+		{"delete expired dashboard versions", srv.deleteExpiredDashboardVersions},
+		{"delete expired images", srv.deleteExpiredImages},
+		{"cleanup old annotations", srv.cleanUpOldAnnotations},
+		{"expire old user invites", srv.expireOldUserInvites},
+		{"delete stale short URLs", srv.deleteStaleShortURLs},
+		{"delete stale query history", srv.deleteStaleQueryHistory},
+		{"delete old login attempts", srv.deleteOldLoginAttempts},
+	}
+
+	logger.Debug("Starting cleanup jobs", "jobs", fmt.Sprintf("%v", cleanupJobs))
+
+	for _, j := range cleanupJobs {
+		if ctx.Err() != nil {
+			logger.Error("Cancelled cleanup job", "error", ctx.Err(), "duration", time.Since(start))
+			return
+		}
+		ctx, span := srv.tracer.Start(ctx, j.name)
+		j.fn(ctx, logger)
+		span.End()
+	}
+
+	logger.Info("Completed cleanup jobs", "duration", time.Since(start))
+}
+
+func (srv *CleanUpService) cleanUpOldAnnotations(ctx context.Context, logger log.Logger) {
 	cleaner := annotations.GetAnnotationCleaner()
 	affected, affectedTags, err := cleaner.CleanAnnotations(ctx, srv.Cfg)
 	if err != nil && !errors.Is(err, context.DeadlineExceeded) {
-		srv.log.Error("failed to clean up old annotations", "error", err)
+		logger.Error("failed to clean up old annotations", "error", err)
 	} else {
-		srv.log.Debug("Deleted excess annotations", "annotations affected", affected, "annotation tags affected", affectedTags)
+		logger.Debug("Deleted excess annotations", "annotations affected", affected, "annotation tags affected", affectedTags)
 	}
 }
 
-func (srv *CleanUpService) cleanUpTmpFiles() {
+func (srv *CleanUpService) cleanUpTmpFiles(ctx context.Context, logger log.Logger) {
 	folders := []string{
 		srv.Cfg.ImagesDir,
 		srv.Cfg.CSVsDir,
 	}
 
 	for _, f := range folders {
-		srv.cleanUpTmpFolder(f)
+		ctx, span := srv.tracer.Start(ctx, fmt.Sprintf("removing old temporary items from directory: %s", strconv.Quote(f)))
+		srv.cleanUpTmpFolder(ctx, logger, f)
+		span.End()
 	}
 }
 
-func (srv *CleanUpService) cleanUpTmpFolder(folder string) {
+func (srv *CleanUpService) cleanUpTmpFolder(ctx context.Context, logger log.Logger, folder string) {
```

Comment:
I don't see `ctx` used in this method; is it needed?

Reply:
It's not, but passing context everywhere is kinda nice, and post #55198 we can remove `logger` and keep only context.Context 🙂

Reply:
It's merged :tada:

Reply:
Thanks! Update this to use `FromContext` instead of passing around loggers. Nice work 🙇